File: Problems/212_wordSearchII.ts
class Trie {
    entry: { [k: string] : any } = {};

    constructor() {
        this.entry = {};
    }

    insert(word: string): void {
        word = word + "#"; //END character
        let nextPossibleLetters = this.entry;        

        for (const letter of word) {
            //check if letter is a next possible path from the current node
            if (nextPossibleLetters[letter] == undefined) {
                //this path doesn't exist yet, so let's create it
                nextPossibleLetters[letter] = {};
            }

            nextPossibleLetters = nextPossibleLetters[letter];
        }
    }

    startsWith(prefix: string): boolean {
        let nextPossibleLetters = this.entry;
        
        for (const letter of prefix) {
            //check if letter is a possible path; if not then the prefix doesn't exist in the Trie, so return false
            if (nextPossibleLetters[letter] == undefined) {
                return false;
            }
            nextPossibleLetters = nextPossibleLetters[letter]
        }

        return true;
    }

    search(word: string): boolean {
        return this.startsWith(word + "#"); //search for entire word with END character
    }
}



function findWords(board: string[][], words: string[]): string[] {
    //use trie; from each position on the board, execute a depth first search: first check if current prefix exists in trie, then check if current prefix exists as a word
    const trie = new Trie();
    for (const word of words) {
        trie.insert(word);
    }

    const wordSet = new Set(words);
    const foundWords: string[] = [];

    const dfs = (currentWord: string, visited: Set<string>, currentRow: number, currentCol: number) => {
        if (visited.has(String([currentRow, currentCol]))) {
            return;
        }

        if (currentRow < 0 || currentRow >= board.length || currentCol < 0 || currentCol >= board[0].length) {
            return;
        }

        visited.add(String([currentRow, currentCol]));
        currentWord += board[currentRow][currentCol];

        if (trie.startsWith(currentWord)) {
            //current word is a prefix in the trie; now we need to check if it is a full word that we haven't found before
            if (wordSet.has(currentWord) && trie.search(currentWord) == true) {
                foundWords.push(currentWord);
                wordSet.delete(currentWord);
            }

            //continue dfs regardless if word was a full word or not
            dfs(currentWord, visited, currentRow - 1, currentCol);
            dfs(currentWord, visited, currentRow + 1, currentCol);
            dfs(currentWord, visited, currentRow, currentCol - 1);
            dfs(currentWord, visited, currentRow, currentCol + 1);
        }
        
        //no point continuing if there are no matching words in trie; before we return we remove our tracks
        visited.delete(String([currentRow, currentCol]))
    }

    for (let row = 0; row < board.length; row += 1) {
        for (let col = 0; col < board[row].length; col += 1) {
            dfs("", new Set(), row, col);
        }
    }

    return foundWords;
};

File: Problems/212_wordSearchII.py
class TrieNode(object):
    def __init__(self):
        self.next = {}
        self.isLeafNode = "" #instead of using true and false, we will use "" or the actual word; this saves us from having to store the word separately in the dfs call

class Solution(object):
    def findWords(self, board, words):
        """
        :type board: List[List[str]]
        :type words: List[str]
        :rtype: List[str]
        """
        
        #backtracking + trie
        foundWords = []

        #first initialise trie as dictionary with words to search for
        root = TrieNode()

        for word in words:
            current = root
            for letter in word:
                if letter not in current.next:
                    current.next[letter] = TrieNode()
                current = current.next[letter]

            current.isLeafNode = word

        #useful constants
        boardHeight = len(board)
        boardWidth = len(board[0])

        visitedSquares = set() #keeps track of all squares visited in the current depth first search
        def SearchSquare(node, row, col):
            #check if we are in a valid row and col
            if row < 0 or row >= boardHeight or col < 0 or col >= boardWidth or (row, col) in visitedSquares:
                return

            visitedSquares.add((row, col))

            #we take in the previous node of the tree to check whether the current letter is a possible next path - if it isn't then the current search doesn't have any matches so we cancel
            letter = board[row][col]
            if letter in node.next:
                lettersNode = node.next[letter]
                #this could be the final letter of the path; i.e. we've found a word
                if lettersNode.isLeafNode != "":
                    foundWords.append(lettersNode.isLeafNode)
                    lettersNode.isLeafNode = ""

                    #todo: if letter node doesn't contain any nodes underneath it, then prune it to effectively remove this word from the trie

                #the current square is also prefix within the trie; so we continue the search
                SearchSquare(lettersNode, row + 1, col)
                SearchSquare(lettersNode, row - 1, col)
                SearchSquare(lettersNode, row, col + 1)
                SearchSquare(lettersNode, row, col - 1)

            visitedSquares.remove((row, col)) #once all calls from current square are complete, we 'retreat'

        for row in range(boardHeight):
            for col in range(boardWidth):
                SearchSquare(root, row, col)

        return foundWords